arm7/lib/src/SND_seq.c

```diff
@@ -0,0 +1,1323 @@
+    {
+        track->channelLLHead = chn->channelLLNext;
+    }
+    else
+    {
+        cur = track->channelLLHead;
+
+        while (cur->channelLLNext)
+        {
+            if (cur->channelLLNext == chn)
+            {
+                cur->channelLLNext = chn->channelLLNext;
+                return;
+            }
+            cur = cur->channelLLNext;
+        }
+    }
+}
+
+static void TrackUpdateChannel(struct SNDTrack *track, struct SNDPlayer *player, int release)
+{
+    int vol;
+    int fader;
+    int pan;
+    int pitch;
+
+    vol = SNDi_DecibelSquareTable[track->volume] + SNDi_DecibelSquareTable[track->expression] +
+          SNDi_DecibelSquareTable[player->volume];
+
+    fader = track->extFader + player->extFader;
+
+    pitch = track->pitchBend;
+    pitch *= track->bendRange << 6;
+    pitch >>= 7;
+    pitch += track->extPitch;
+
+    pan = track->pan;
+
+    if (track->panRange != 127)
+        pan = (pan * track->panRange + 0x40) >> 7;
+
+    pan += track->extPan;
+
+    if (vol < -0x8000)
+        vol = -0x8000;
+
+    if (fader < -0x8000)
+        fader = -0x8000;
+
+    if (pan < -128)
+        pan = -128;
+    else if (pan > 127)
+        pan = 127;
+
+    for (struct SNDExChannel *chn = track->channelLLHead; chn != NULL; chn = chn->channelLLNext)
+    {
+        chn->userDecay2 = (short)fader;
+
+        if (chn->envStatus == 3)
+            continue;
+
+        chn->userDecay = (s16)vol;
+        chn->userPitch = (s16)pitch;
+        chn->userPan = (s8)pan;
+        chn->panRange = track->panRange;
+        chn->lfo.param = track->mod;
+
+        if (chn->length == 0 && release != 0)
+        {
+            chn->priority = 1;
+            SND_ReleaseExChannel(chn);
+        }
+    }
+}
+
+static void PlayerUpdateChannel(struct SNDPlayer *player)
+{
+    for (int i = 0; i < SND_TRACK_COUNT_PER_PLAYER; i++)
+    {
+        struct SNDTrack *trk = PlayerGetTrack(player, i);
+        if (trk)
+        {
+            TrackUpdateChannel(trk, player, 1);
+        }
+    }
+}
+
+static void TrackPlayNote(
+    struct SNDTrack *track, struct SNDPlayer *player, int midiKey, int velocity, int length)
+{
+    struct SNDExChannel *chn = NULL;
+
+    if (track->flags.tie)
+    {
+        chn = track->channelLLHead;
+        if (chn)
+        {
+            chn->midiKey = (u8)midiKey;
+            chn->velocity = (u8)velocity;
+        }
+    }
+
+    if (chn == NULL)
+    {
+        struct SNDInstData inst;
+        if (!SND_ReadInstData(player->bank, track->program, midiKey, &inst))
+            return;
+
+        u32 allowedChannels;
+
+        switch (inst.type)
+        {
+            case SND_INST_PCM:
+            case SND_INST_DIRECTPCM:
+                // all channels support PCM
+                allowedChannels = 0xFFFF;
+                break;
+            case SND_INST_PSG:
+                // only channels 8, 9, 10, 11, 12, 13 support PSG
+                allowedChannels = 0x3F00;
```

Comment:
this feels like there should be constants here

Reply:
Not really, this is just a bit mask with the bits set which are mentioned in the comment. No further meaning behind that value.